File: old/actions/install_docker.py
"""
Install Docker action
"""
import logging
from cli import Apt, Docker, Command, Curl, Shell
from .base import Action
logger = logging.getLogger(__name__)

class InstallDockerAction(Action):
    """Action to install Docker"""
    description = "docker installation"

    def execute(self) -> bool:
        """Install Docker"""
        if not self.ssh_service:
            logger.error("SSH service not initialized")
            return False

        logger.info("Installing Docker...")

        # Check if curl is available
        curl_check_output, curl_check_exit = self.ssh_service.execute(
            Command().command("curl").exists(), 
            sudo=False
        )

        has_curl = curl_check_exit == 0 and curl_check_output and "curl" in curl_check_output
        if has_curl:
            # Try official Docker install script
            logger.info("Downloading Docker install script...")

            download_cmd = Curl().fail_silently(True).silent(True).show_errors(True).location(True).output("/tmp/get-docker.sh").url("https://get.docker.com").download()
            download_output, download_exit = self.ssh_service.execute(download_cmd, sudo=False)

            if download_exit is not None and download_exit != 0:
                logger.warning("Failed to download Docker install script: %s", download_output)
                has_curl = False
            else:
                logger.info("Running Docker install script...")
                
                script_cmd = Shell().shell("sh").script("/tmp/get-docker.sh").execute()
                script_output, script_exit = self.ssh_service.execute(script_cmd, sudo=False, timeout=300)

                # Check for installation errors in output
                if script_output and ("E: Package" in script_output or "Unable to locate package" in script_output or "has no installation candidate" in script_output):

                    logger.warning("Docker install script failed, falling back to docker.io")

                    has_curl = False
                elif script_exit is not None and script_exit != 0:

                    logger.warning("Docker install script failed with exit code %s, falling back to docker.io", script_exit)

                    has_curl = False
                else:
                    # Verify Docker is installed after script
                    docker_check_cmd = Docker().is_installed_check()
                    check_output, check_exit_code = self.ssh_service.execute(docker_check_cmd, sudo=True)

                    if Docker.parse_is_installed(check_output):

                        logger.info("Docker installed successfully via install script")

                        return True

                    logger.warning("Docker install script completed but Docker not found, falling back to docker.io")

                    has_curl = False

        # Fallback to docker.io via apt
        if not has_curl:

            logger.info("Installing docker.io via apt...")

            install_cmd = Apt().install(["docker.io"])

            if self.apt_service:
                install_output = self.apt_service.execute(install_cmd, timeout=300)

                # APTService.execute returns output string, check for errors
                if install_output and ("E: Package" in install_output or "Unable to locate package" in install_output or "has no installation candidate" in install_output):

                    logger.error("Docker installation failed - packages not found")
                    logger.error("Docker installation output: %s", install_output[-1000:])

                    return False
            else:
                install_output, install_exit = self.ssh_service.execute(f"sudo -n {install_cmd}", sudo=False, timeout=300)

                if install_exit is not None and install_exit != 0:
                    logger.error("Docker installation failed with exit code %s", install_exit)

                    if install_output:
                        logger.error("Docker installation output: %s", install_output[-1000:])
                    return False

                if install_output and ("E: Package" in install_output or "Unable to locate package" in install_output or "has no installation candidate" in install_output):

                    logger.error("Docker installation failed - packages not found")
                    logger.error("Docker installation output: %s", install_output[-1000:])

                    return False

        # Verify Docker is installed
        docker_check_cmd = Docker().is_installed_check()
        check_output, check_exit_code = self.ssh_service.execute(docker_check_cmd, sudo=True)

        if not Docker.parse_is_installed(check_output):

            logger.error("Docker installation failed - verification shows Docker is not installed")

            return False

        logger.info("Docker installed successfully")

        return True

File: src/actions/install_docker.go
package actions

import (
	"fmt"
	"strings"
	"enva/cli"
	"enva/libs"
	"enva/services"
)

// InstallDockerAction installs Docker
type InstallDockerAction struct {
	*BaseAction
}

func NewInstallDockerAction(sshService *services.SSHService, aptService *services.APTService, pctService *services.PCTService, containerID *string, cfg *libs.LabConfig, containerCfg *libs.ContainerConfig) Action {
	return &InstallDockerAction{
		BaseAction: &BaseAction{
			SSHService:   sshService,
			APTService:   aptService,
			PCTService:   pctService,
			ContainerID: containerID,
			Cfg:          cfg,
			ContainerCfg: containerCfg,
		},
	}
}

func (a *InstallDockerAction) Description() string {
	return "docker installation"
}

func (a *InstallDockerAction) Execute() bool {
	if a.SSHService == nil {
		libs.GetLogger("install_docker").Printf("SSH service not initialized")
		return false
	}
	libs.GetLogger("install_docker").Printf("Installing Docker...")
	
	// Check if curl is available
	curlCheckOutput, curlCheckExit := a.SSHService.Execute(cli.NewCommand().SetCommand("curl").Exists(), nil)
	hasCurl := curlCheckExit != nil && *curlCheckExit == 0 && strings.Contains(curlCheckOutput, "curl")
	
	if hasCurl {
		libs.GetLogger("install_docker").Printf("Downloading Docker install script...")
		downloadCmd := cli.NewCurl().FailSilently(true).Silent(true).ShowErrors(true).Location(true).Output("/tmp/get-docker.sh").URL("https://get.docker.com").Download()
		downloadOutput, downloadExit := a.SSHService.Execute(downloadCmd, nil)
		
		if downloadExit == nil || *downloadExit != 0 {
			libs.GetLogger("install_docker").Printf("Failed to download Docker install script: %s", downloadOutput)
			hasCurl = false
		} else {
			libs.GetLogger("install_docker").Printf("Running Docker install script...")
			scriptCmd := cli.NewShell().Script("/tmp/get-docker.sh").Execute()
			timeout := 300
			scriptOutput, scriptExit := a.SSHService.Execute(scriptCmd, &timeout)
			
			if strings.Contains(scriptOutput, "E: Package") || strings.Contains(scriptOutput, "Unable to locate package") || strings.Contains(scriptOutput, "has no installation candidate") {
				libs.GetLogger("install_docker").Printf("Docker install script failed, falling back to docker.io")
				hasCurl = false
			} else if scriptExit != nil && *scriptExit != 0 {
				libs.GetLogger("install_docker").Printf("Docker install script failed with exit code %d, falling back to docker.io", *scriptExit)
				hasCurl = false
			} else {
				dockerCheckCmd := cli.NewDocker().IsInstalledCheck()
				checkOutput, _ := a.SSHService.Execute(dockerCheckCmd, nil)
				if cli.ParseDockerIsInstalled(checkOutput) {
					libs.GetLogger("install_docker").Printf("Docker installed successfully via install script")
					return true
				}
				libs.GetLogger("install_docker").Printf("Docker install script completed but Docker not found, falling back to docker.io")
				hasCurl = false
			}
		}
	}
	
	// Fallback to docker.io via apt
	if !hasCurl {
		libs.GetLogger("install_docker").Printf("Installing docker.io via apt...")
		installCmd := cli.NewApt().Install([]string{"docker.io"})
		if a.APTService != nil {
			installOutput, exitCode := a.APTService.Install([]string{"docker.io"})
			if exitCode == nil || *exitCode != 0 {
				if strings.Contains(installOutput, "E: Package") || strings.Contains(installOutput, "Unable to locate package") || strings.Contains(installOutput, "has no installation candidate") {
					libs.GetLogger("install_docker").Printf("Docker installation failed - packages not found")
					outputLen := len(installOutput)
					start := 0
					if outputLen > 1000 {
						start = outputLen - 1000
					}
					libs.GetLogger("install_docker").Printf("Docker installation output: %s", installOutput[start:])
					return false
				}
			}
		} else {
			_, _ = a.SSHService.Execute(fmt.Sprintf("sudo -n %s", installCmd), nil)
			timeout := 300
			installOutput2, installExit2 := a.SSHService.Execute(installCmd, &timeout)
			if installExit2 != nil && *installExit2 != 0 {
				libs.GetLogger("install_docker").Printf("Docker installation failed with exit code %d", *installExit2)
				outputLen := len(installOutput2)
				start := 0
				if outputLen > 1000 {
					start = outputLen - 1000
				}
				libs.GetLogger("install_docker").Printf("Docker installation output: %s", installOutput2[start:])
				return false
			}
			if strings.Contains(installOutput2, "E: Package") || strings.Contains(installOutput2, "Unable to locate package") || strings.Contains(installOutput2, "has no installation candidate") {
				libs.GetLogger("install_docker").Printf("Docker installation failed - packages not found")
				outputLen := len(installOutput2)
				start := 0
				if outputLen > 1000 {
					start = outputLen - 1000
				}
				libs.GetLogger("install_docker").Printf("Docker installation output: %s", installOutput2[start:])
				return false
			}
		}
	}
	
	// Verify Docker is installed
	dockerCheckCmd := cli.NewDocker().IsInstalledCheck()
	checkOutput, _ := a.SSHService.Execute(dockerCheckCmd, nil)
	if !cli.ParseDockerIsInstalled(checkOutput) {
		libs.GetLogger("install_docker").Printf("Docker installation failed - verification shows Docker is not installed")
		return false
	}
	libs.GetLogger("install_docker").Printf("Docker installed successfully")
	return true
}

